### PR TITLE
🪟 Per-Stream state new flow

### LINKED ISSUE #14560: Per-Stream adjust frontend flow
As part of the per-stream process we discovered some shortcoming last week and will require a slightly different slow than we currently have. The full flow for the FE what happens is documented in [this flowchart](https://whimsical.com/per-stream-updates-2022-07-06-5t1ecmGWPhLm9xoRU8rZru).

The main adjustments to make in the FE are:

When the user hits the "save" button when updating a connection, we need to check, if the configured catalog we're about to send to the `web_backend/connection/update` API differs anyhow (deep compare) from the initial catalog we loaded. If we "refreshed the source catalog" in the meantime, we still need to compare against the original one that was loaded when the page was loaded (not the one we got back from the `web_backend/connection/read` (`withRefreshedCatalog: true`) call.

If they are still the same, we simply call `web_backend/connection/update` as we do today. We're not needing to send `withRefreshedCatalog` to that API, since we're getting rid of it.

If the catalogs differ anyhow, we'll make a call to `web_backend/state/get_type` first to determine the state type of the connection. After that we show a modal warning the user, that we'll reset their streams. The wording depends on the state type we received from the prior API call. For "legacy" a full reset will happen, so we mention this, for "global" and "stream" only changed streams will be resetted, we'll mention that in the warning modal (exact wording still to be done on the PR). The modal does not offer an option not to reset (as our current one does). It just will allow confirming or cancelling (in which case the connection won't be saved). If the user confirms, we'll call `web_backend/connection/update` (also without `withRefreshedCatalog`) with the new changes.

We'll be getting rid of the current modal that asks whether or not the connection should be reset or not.

All APIs are already available. We can create that PR, but must wait with merging for the corresponding API changes to be ready, so that the behavior will match our expectations.

### PR DESCRIPTION
**Pull request by timroes**
## What

Fixes #14560

**Partial reset possible:**
screenshot-20220713-114036

**Full reset required:**
screenshot-20220713-114300

#### TODOs
- [ ] ~Unit tests~
- [x] Proper typings for ModalService input/output
- [x] Finish and cleanup modal service
- [x] Proper warning modal
- [x] Finalize wording
- [x] Remove secondaryButton from ConfirmationModal again
- [x] Switch to new update API once merged

### CODE REVIEW

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (timroes) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:118]: :information_source: The `onSubmit` function was never called with the `formikHelpers` parameters specified, so this should have never been called.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (lmossman) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:142]: I think we probably want this to set `skipReset` to `true` in this case. The reason being that we won't ask for any confirmation, so we don't want to accidentally cause a reset to happen if, for example, another user changed the configuration as this request is being sent to the backend. If `skipReset` was false, that scenario would result in use calculating the diff and finding a change, and submitting a reset, even though we never prompted the user for confirmation of this.

Setting `skipReset` to `true` here guarantees that when we don't ask the user for confirmation, then that save will not cause a reset to be submitted

Comment (timroes) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:142]: Thanks. This would also be my preferred way, to never accidentally reset data.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (timroes) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:173]: :information_source: This solves some issue that the "This connection was saved" reappears again after an error vanished, despite no new save has happened.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx
Comment (timroes) [inline airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:169]: :information_source: This was the old "should we reset or not" modal, that does not longer exist with this change.

**Review by lmossman — COMMENTED** [review]
These changes make sense to me! Thanks for making the loom video to walk through the changes, that was really helpful for understanding what is going on here.

The only reason I am not approving this PR is because there is still a couple of backend changes that need to be made the the `/updateNew` endpoint to have it properly perform a reset if only some stream configuration (not schema) was changed, e.g. just the sync mode was changed. Because currently that endpoint will not detect config-only changes and will not perform a reset in that case.

Those PRs are:
https://github.com/airbytehq/airbyte/pull/14603
https://github.com/airbytehq/airbyte/pull/14626
I will try to work with Benoit to get those merged in as soon as we can. Once they are merged, I will approve this PR and we can make this change

**Comment by Amruta-Ranade** [comment]
Text suggestion: 

Changing the stream configuration will reset (delete and resync data) streams, and the affected data will be unavailable in the destination until the next sync completion. Resetting is still a safer option since skipping a reset can cause a failed sync or inconsistent data.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (edmundito) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:58]: Not sure if it's clear that `onClose` confirms the action. I would it expect that onClose just closes without consequences.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:166]: alternatively:
```ts
if (result?.submitCancelled) return
```

File: airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:120]: `canSubmitUntouchedForm`? This component has no use of the word "save" but instead it cares about submit

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/EditControls.tsx
Comment (edmundito) [inline airbyte-webapp/src/views/Connection/ConnectionForm/components/EditControls.tsx:14]: Same behavior as `disabled`, maybe should be called `disabled`

**Review by edmundito — COMMENTED** [review]
I reviewed the whimiscal and followed along. The implementation looks good so far. I have some comments here and there and waiting to consolidate with the current ModalBody/ModalFooter changes.

File: airbyte-webapp/src/core/domain/connection/utils.ts
Comment (lmossman) [reply airbyte-webapp/src/core/domain/connection/utils.ts:22]: That's correct, it defaults to `false` in the backend if it is not present

**Comment by lmossman** [comment]
Both of the PRs I mentioned in my previous comment have been merged into the `updateNew` endpoint, so that should now properly execute resets when only the stream configuration is changed (if the skipReset flag is false of course), so if you rebase this PR on master you should see that correct behavior when testing this PR

File: airbyte-webapp/src/core/domain/connection/WebBackendConnectionService.ts
Comment (lmossman) [reply airbyte-webapp/src/core/domain/connection/WebBackendConnectionService.ts:7]: Yes it is; this endpoint was created so that we could merge backend changes related to this new flow in before the frontend changes were merged. We have a ticket here to remove this temporary endpoint after the launch of per-stream: https://github.com/airbytehq/airbyte/issues/14733

File: airbyte-webapp/src/hooks/services/useConnectionHook.tsx
Comment (timroes) [reply airbyte-webapp/src/hooks/services/useConnectionHook.tsx:197]: Not really. This was in there since 2 years, so I fear that knowledge is lost. Looking at the old code it seems this was a separate parameter to this hook earlier, and thus spreaded like this into it (trying to preventing actually setting the value when it was undefined), but it's really just guessing.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (timroes) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:118]: It seems to work. This should all be handled by the call inside the `ConnectionForm`: https://github.com/airbytehq/airbyte/blob/tim/per-stream-flow/airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx#L167

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (timroes) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:166]: That message was just another hint about your data will being cleaned when pressing save. Due to the warning modal it's now obvious bevore saving, so I decided to remove it (since it also looked very "out of place" in terms of design).

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (timroes) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:192]: The logic for this is nested a bit deeper: https://github.com/airbytehq/airbyte/blob/master/airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx#L213-L217 In readonly mode the additionalSchemaControl will never render out. Is this the right place to do this... most likely not, given that the child component should have no knowledge what this `additionalSchemaControl` actually is, so should likely be refactored during our Connection page cleanup (cc @tealjulia )

File: airbyte-webapp/src/components/Modal/ModalBody.module.scss
Comment (timroes) [inline airbyte-webapp/src/components/Modal/ModalBody.module.scss:5]: :information_source: @edmundito I've changed this to `auto`, since `scroll` will have scrollbars always visible (on non mac systems) inside the modal body:

screenshot-20220715-111421

File: airbyte-webapp/src/components/Modal/Modal.tsx
Comment (timroes) [inline airbyte-webapp/src/components/Modal/Modal.tsx:26]: :information_source: Changed this to use `key` instead of the deprecated `keyCode`, so I could test it via `@testing-library/user-events`.

File: airbyte-webapp/src/components/Modal/ModalBody.module.scss
Comment (edmundito) [reply airbyte-webapp/src/components/Modal/ModalBody.module.scss:5]: Whoops, yes I am familiar with this. it definitely should be auto!

**Review by edmundito — COMMENTED** [review]
Code looks good

**Comment by edmundito** [comment]
Code is good, but I ran into a situation while testing:

1. Add a new table to the source
2. Open the replication settings
3. Refresh the schema
4. Hit save without changing the stream configuration (new table is unchecked)

I didn't change anything about the streams, but I still got the dialog to reset the schema.

**Comment by edmundito** [comment]
Another thing I noticed: After adding the new table and saving the schema, the stream disappears from the list. But when I reload the replication page, it appears again.

**Comment by edmundito** [comment]
Another thing about the modal, for me it makes more sense that it shows the "Save and Reset" as a button rather than a checkbox. so the user has three CTAs:
1. Cancel
2. Save
3. Save and Reset streams

**Comment by timroes** [comment]
> Another thing about the modal, for me it makes more sense that it shows the "Save and Reset" as a button rather than a checkbox. so the user has three CTAs:
> 1. Cancel
> 2. Save
> 3. Save and Reset streams

The problem is we really want to discourage saving without resetting, thus this layout. Cc @andyjih happy to get your input on that question, checkbox vs three buttons?

**Comment by timroes** [comment]
> Code is good, but I ran into a situation while testing:
> 
> 1. Add a new table to the source
> 2. Open the replication settings
> 3. Refresh the schema
> 4. Hit save without changing the stream configuration (new table is unchecked)
> 
> I didn't change anything about the streams, but I still got the dialog to reset the schema. This also happens if I modify a disabled stream.

That is so far actually intended behavior, since the catalog changed (despite the stream not being enabled that changed).

@andyjih @lmossman I wonder if we should actually only compare enabled streams in the original and the catalog we're about to send? And only if there's a difference in the enabled ones show that modal?

**Comment by lmossman** [comment]
> That is so far actually intended behavior, since the catalog changed (despite the stream not being enabled that changed).
> 
> @andyjih @lmossman I wonder if we should actually only compare enabled streams in the original and the catalog we're about to send? And only if there's a difference in the enabled ones show that modal?

@timroes I believe this is actually how it already works in the backend - we only compare the configured catalog that the frontend passes us with the configured catalog that is stored in the database, and those configured catalogs only contain streams that are enabled. So I think it will be a purely frontend change to bring it in line with this behavior, which we should do so that they are consistent

**Comment by lmossman** [comment]
@timroegs you should rebase or merge master again, as on Friday I merged a fix for a bug we found in the updateNew endpoint: https://github.com/airbytehq/airbyte/pull/14763

**Comment by timroes** [comment]
> , and those configured catalogs only contain streams that are enabled. 

Just for double checking: At the moment (before and with this PR) we're sending to the backend also all non-enabled streams, and have `selected: false` set on them. So with the change of only comparing active streams, we'd compare on a subset of what we're sending to the backend. This would still though be the desired behavior?

**Comment by timroes** [comment]
Added the change to only compare selected streams for showing the modal.

**Comment by timroes** [comment]
@edmundito Could you give this another look? With Andy out, Charles and I would prefer merging the changes (once BE is fixed) for now as it is, and do a separate PR for cleaning up the modal (wording + design) once Andy will be available again.

**Review by andyjih — CHANGES_REQUESTED** [review]

**Review by edmundito — APPROVED** [review]
Code changes look good. Will approve pending @andyjih's comments on the wording for final approval.

**Review by andyjih — APPROVED** [review]

**Pull request merged**
